feat: (network) increase abstraction level of the connection classes …

ptesavol opened the pull request:
Increase abstraction level of the connection classes to allow for browser implementation

ptesavol committed:
feat: (network) increase abstraction level of the connection classes to allow for browser implementation

harbu commented on packages/network/src/connection/NodeWebRtcEndpoint.ts:
```diff
@@ -0,0 +1,43 @@
+import { NodeWebRtcConnection } from './NodeWebRtcConnection'
+import { WebRtcConnection } from './WebRtcConnection'
+
+export class NodeWebRtcEndpoint extends WebRtcEndpoint implements IWebRtcEndpoint {
```
Wondering whether we should favor composition over inheritance here? The only two things needed by `WebRtcEndpoint` is a way to create new connections `doCreateConnection` and a clean up method `doStop`. We could create a factory interface that we pass to `WebRtcEndpoint` which it uses to do the two things avoiding inheritance altogether.

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+// reminder: only use Connection emitter for external handlers
+// to make it safe for consumers to call removeAllListeners
+// i.e. no this.on('event')
+export const ConnectionEmitter = EventEmitter as { new(): StrictEventEmitter<EventEmitter, Events> }
```
Duplicated code segment from `WebRtcConnection`, can be removed?

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+        this.onStateChange = this.onStateChange.bind(this)
+        this.onLocalCandidate = this.onLocalCandidate.bind(this)
+        this.onLocalDescription = this.onLocalDescription.bind(this)
+        this.onStateChange = this.onStateChange.bind(this)
```
onStateChange already binded above?
```suggestion
```

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+    connect(): void {
+        if (this.isFinished) {
+            throw new Error('Connection already closed.')
+        }
```
The problem that I see here is that `this.isFinished` is a field of the super class. This means that every implementer of a subclass must remember to do this check in the beginning of their `connect()` implementation, which ofc, is very easy to forget to do and its duplicated logic.

Some possible solutions:
1. Make each sub class keep track of their own `isFinished`
2. Make `connect` non-abstract, implement it in the super class, and then provide a new protected method `doConnect` that actually does the connection,i .e.

In (super) class `WebRtcConnection`:
```ts
connect(): void {
    if (this.isFinished) {
        throw new Error('Connection already closed.')
    }
    this.doConnect()
}

...

protected abstract doConnect(): void
```

harbu commented on packages/network/src/connection/WebRtcConnection.ts:
```diff
@@ -0,0 +1,402 @@
+    abstract getBufferedAmount(): number 
+    abstract getMaxMessageSize(): number 
+    abstract isOpen(): boolean
+    protected abstract doSendMessage(message: string): boolean
```
Might want to write a JSDoc comment on this method. What does the return value indicate? What happens when it is true or false? Also what guarantees do we have about state when we invoke `doSendMessage`? For example, in `NodeWebRtcConnection` the implementation is as follows

```ts
protected doSendMessage(message: string): boolean {
    return this.dataChannel!.sendMessage(message)
}
```
Clearly there is an assumption that `isOpen() === true` holds for an instance  before `doSendMessage` is ever invoked. This may be worth commenting, e.g., "connectivity is checked before base class invokes `doSendMessage`, i.e., checking `isOpen()`  is done by base class".

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+            this.connectionEmitter.on('dataChannel', this.onDataChannel)
+        }
+
+        this.connectionTimeoutRef = setTimeout(() => {
```
It's not a good idea to be setting directly a super class's field. Need to decide whether timeout logic should be _shared_ or _subclass-specific_.

1. If subclass-specific, then just move the field up from `WebRtcConnection` up to`NodeWebRtcConnection`.
2. If shared, then make `connectionTimeoutRef` private in `WebRtcConnection` and have it handle the whole thing (e.g. in a `connect` method as I described in previous comment). If the timeout occurs, I believe it will call abstract `close` anyways?

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+            try {
+                this.connection.setRemoteDescription(description, type)
+            } catch (err) {
+                this.logger.warn('setRemoteDescription failed, reason: %s', err)
```
Wondering if we should have a separate logger for subclasses just for clarity? E.g. we know which sort of subclass we are looking at based on the log message.

harbu replied:
I can't use the same variable name `logger` in the subclass even though the variable with the same name is private in the base class. So just gonna skip this, because it feels funny to have to name the logger in the subclass as something else 🤷

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+    }
+
+    close(err?: Error): void {
+        if (this.isFinished) {
```
Same thing as before, this should really be checked by the base class not here.

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+            return
+        }
+
+        this.isFinished = true
```
Yeah this should be base class's responsibility, we should not really set values from subclass.

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+        this.dataChannel = null
+        this.connection = null
+
+        this.doClose()
```
I really think it should be the other way around here.

In `WebRtcConnection` we should have the public `close` method and an abstract `doClose` method implemented by subclasses, e.g.
```ts
close(err?: Error): void {
    if (this.isFinished) {
       return
    }
    this.isFinished = true
    ....
    this.doClose()
}

...

protected doClose(err?: Error): void
```

And then the method here in this `NodeWebRtcConnection` would be named `doClose` instead.

This way the implementer does not have to remember or know that they need to call `doClose` at the end of their `close` implementation. Instead it is all taken care of by the superclass!

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+    private onStateChange(state: string): void {
+        this.logger.trace('conn.onStateChange: %s -> %s', this.lastState, state)
+
+        this.lastState = state
```
It's a bit cumbersome to have to remember as an implementer of a sub class to set this value every time a state changes. My recommended change would be the following.

- Remove field `this.lastState` from `WebRtcConnection`
- Add a new abstract method (to replace the above) into `WebRtcConnection`, e.g. `protected abstract getLastState(): string`
- `WebRtcConnection` should solely rely on method `getLastState()` to get the last state
- Add field `this.lastState` to `NodeWebRtcConnection` and implement abstract method `getLastState()` so that it returns this value

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+                if (this.isFinished) { return }
+                this.logger.warn(`connection timed out after ${this.newConnectionTimeout}ms`)
+                this.close(new Error(`timed out after ${this.newConnectionTimeout}ms`))
+            }, this.newConnectionTimeout)
```
If the plan is to have a shared timeout logic within `WebRtcEndpoint` class, then my proposal would be to remove this with a method call to `WebRtcEndpoint`, e.g.

In WebRtcEndpoint:
```ts
protected resetConnectionTimeout(): void {
            clearTimeout(this.connectionTimeoutRef!)
            this.connectionTimeoutRef = setTimeout(() => {
                if (this.isFinished) { return }
                this.logger.warn(`connection timed out after ${this.newConnectionTimeout}ms`)
                this.close(new Error(`timed out after ${this.newConnectionTimeout}ms`))
            }, this.newConnectionTimeout)
}
```

And then replace the code here as follows:
```ts
        if (state === 'disconnected' || state === 'closed') {
            this.close()
        } else if (state === 'failed') {
            this.close(new Error('connection failed'))
        } else if (state === 'connecting') {
            this.resetConnectionTimeout()
        }
```

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+
+    private onGatheringStateChange(state: string): void {
+        this.logger.trace('conn.onGatheringStateChange: %s -> %s', this.lastGatheringState, state)
+        this.lastGatheringState = state
```
Same idea as with `lastState` above.

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+    }
+
+    private onLocalDescription(description: string, type: DescriptionType): void {
+        this.emit('localDescription', type, description)
```
Let's consider moving the event emission to base class `WebRtcConnection`.

I.e. replace this with
```ts
private onLocalDescription(description: string, type: DescriptionType): void {
        super.onLocalDescription(type, description) // super equivalent?
}
```

And then in the base class `WebRtcConnection`:

```ts
protected onLocalDescription(description: string, type: DescriptionType): void {
    this.emit('localDescription', type, description)
}
```

Additionally we need to add a comment to the class JSDoc of `WebRtcEndpoint` that says that "implementers of sub classes need to take care to call protected method onLocalDescription whenever their underlying webrtc library generates a local description".

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+    }
+
+    private onLocalCandidate(candidate: string, mid: string): void {
+        this.emit('localCandidate', candidate, mid)
```
Same thinking as previous comment.

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+    }
+
+    private setupDataChannel(dataChannel: DataChannel): void {
+        this.paused = false
```
Again, we should look into moving the backpressure pausing logic to either be shared or subclass-specific. Shouldn't need to set superclass fields directly.

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+            if (!this.paused) { return }
+            this.paused = false
+            this.setFlushRef()
+            this.emit('bufferLow')
```
Again consider whether buffer handling logic should be shared or subclass-specific.

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+            } else {
+                this.emit('message', msg.toString()) // TODO: what if we get binary?
+            }
+        })
```
The handling of the message here and checking whether ping should probably live under `WebRtcConnection`.

harbu commented on packages/network/src/connection/NodeWebRtcConnection.ts:
```diff
@@ -0,0 +1,318 @@
+    private openDataChannel(dataChannel: DataChannel): void {
+        if (this.connectionTimeoutRef !== null) {
+            clearTimeout(this.connectionTimeoutRef)
+        }
```
Should be done by `WebRtcEndpoint`, e.g. within `emitOpen()`.

harbu commented on packages/network/src/connection/WebRtcConnection.ts:
```diff
@@ -0,0 +1,402 @@
+    protected readonly bufferThresholdLow: number
+    protected pingAttempts = 0
+    protected rtt: number | null
+    protected rttStart: number | null
```
Let's try to make as many as possible to `protected readonly`, if not possible keep `private` and provide getter.

harbu reviewed (changes requested):
See comments; I can also implement them if needed

Commit:
Refactoring suggestions (#111)

* refactor: replace NodeWebRtcEndpoint with composition

* refactor: rm duplicated code segment

* refactor: NodeWebRtcConnection remove duplicate .bind

* refactor: inheritance round #1

- move public method `connect` to base class, add abstract protected
  `doConnect` method as replacement for subclasses
- move public method `close` to base class, add abstract protected
  `doClose` method as replacement for subclasses
- make `connectionTimeoutRef` field private and move all related logic to base class
- make ìsFinished` field private
- add method `restartConnectionTimeout` to base class to allow a subclass to
  forcefully restart a connection timer

* refactor: add JSDOC to method doSendMessage

* refactor: move last and lastGatheringState to subclass

* refactor: add methods to base class to emit

* refactor: make paused field private in base class

- Field `paused` is private in base class
- Add protected method emitLowBackPressure thru which subclass can
  inform base calss of low back pressure
- Therefore backpressure managment is now taken care of by base class
- Made `setFlushRef` private in base class

* refactor: WebRtcConnection#emitMessage for handling message

* refactor: fix accidental lastGatheringState private

* refactor: separate logger for base and sub classes

* refactor: privatize WebRtcConnection fields

* fix: add class-level JSDoc to WebRtcConnection

* fix: make eslint happy

* fix: address PR comment, move factories
